### PR TITLE
[WIP] fetch tinyllama

### PR DESCRIPTION
**Pull request by rodfer0x80**
TLDR:
-- add fetching tinyllama from ~/.cache/tinygrad/downloads to ./weights/Tiny-LLaMA/1B from HuggingFace https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T

Log: 
PYTHONPATH="."  CPU=1 python examples/llama.py --prompt="Hello." --gen="tiny" --size="1B" --count=50 --temperature=0
using CPU backend
using LLaMA-tiny-1B model
https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T/resolve/main/model.safetensors?download
https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T/resolve/main/model.safetensors?download
https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T/resolve/main/model.safetensors?download
https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T/resolve/main/model.safetensors?download
https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T/resolve/main/model.safetensors?download
https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T/resolve/main/model.safetensors?download
https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T/resolve/main/model.safetensors?download
https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T/resolve/main/tokenizer.model?download=t
ram used:  4.40 GB, freqs_cis                                         : 100%|███████| 202/202 [00:01<00:00, 131.50it/s]
loaded weights in 1536.88 ms, 4.40 GB loaded at 2.86 GB/s
Hello.

A: You can use the following code to get the^C

PYTHONPATH="."  CPU=1 python examples/llama.py --prompt="Hello." --gen="tiny" --size="1B" --count=10 --temperature=0
using CPU backend
using LLaMA-tiny-1B model
ram used:  4.40 GB, freqs_cis                                         : 100%|███████| 202/202 [00:01<00:00, 109.57it/s]
loaded weights in 1849.09 ms, 4.40 GB loaded at 2.38 GB/s
Hello.

A: You can use the following code

### CODE REVIEW

**Comment by rodfer0x80** [comment]
tokenizer_path = ./weights/LLaMA-tiny/tokenizer.model
model_path = ./weights/LLaMA-tiny/model.safetensors

**Comment by geohot** [comment]
Can you factorize this to be nice?

**Comment by g1y5x3** [comment]
their chat mode weights are in bf16, only 2.2G. https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v0.6

**Comment by geohot** [comment]
oh we want the 2.2G ones

**Comment by g1y5x3** [comment]
Should be able to call `apply_chat_template` directly from `tokenizer` example from https://huggingface.co/docs/transformers/main/en/chat_templating#how-do-i-use-chat-templates
```
messages = [
    {
        "role": "system",
        "content": "You are a friendly chatbot who always responds in the style of a pirate",
    },
    {"role": "user", "content": "How many helicopters can a human eat in one sitting?"},
 ]
tokenized_chat = tokenizer.apply_chat_template(messages, tokenize=True, add_generation_prompt=True, return_tensors="pt")
print(tokenizer.decode(tokenized_chat[0]))
```

File: examples/llama.py [hunk 1]
```diff
@@ -340,6 +346,21 @@ def greedy_until(self, prompt:str, until, max_length, temperature):
...
+    if not model_safetensors_path.exists():
+      LLaMa._fetch_and_replace_file(model_url, model_safetensors_path)
+    if not tokenizer_path.exists():
+      LLaMa._fetch_and_replace_file(tokenizer_url, tokenizer_path)
+    return model_safetensors_path
+
+  @staticmethod
+  def _fetch_and_replace_file(url, destination_path):
+    downloaded_path = fetch(url)
+    os.replace(downloaded_path, destination_path)
```
Comment (g1y5x3) [inline examples/llama.py:362]: fetch already returns pathlib.Path so you should be able to use replace/rename from pathlib directly without calling os to do the same

**Comment by rodfer0x80** [comment]
Need support for BF16 first, in tinygrad.helpers.safeload function?
Will have a look at it tomorrow, have to get up in a few hours. 

 File "/home/rodfer/workspace/tinycorp/tinygrad/examples/llama.py", line 318, in build
    weights = load(str(model_path))
              ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rodfer/workspace/tinycorp/tinygrad/examples/llama.py", line 247, in load
    return safe_load(fn)
KeyError: 'BF16'

**Comment by g1y5x3** [comment]
With llama torch tensors, currently its being [converted to fp16 manually](https://github.com/tinygrad/tinygrad/blob/8f89e21fcab568a8822ce92a2df8ab0c075164c9/tinygrad/nn/state.py#L75-L80). you can probably apply it to safe tensors

File: examples/llama.py [hunk 2]
```diff
@@ -296,8 +296,14 @@ def quantize(tensors):
     return new_tensors
 
 class LLaMa:
+  TINY_MODEL_URL = "https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v0.6/resolve/main/model.safetensors?download=true"
+  TINY_TOKENIZER_URL = "https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v0.6/resolve/main/tokenizer.model?download=true"
```
Comment (g1y5x3) [inline examples/llama.py:300]: I think these two can put under the dict `MODEL_PARAMS` `tiny` `1B`

**Comment by rodfer0x80** [comment]
I  have tried applying the same casting from the torch_load function but it started outputting gibberish. 
I am not sure what I am missing, they are both uint8 -> bfloat16? Something about the offsets??


PYTHONPATH="."  CPU=1 python examples/llama.py --prompt="Hello." --gen="tiny" --size="1B" --count=10 --temperature=0
using CPU backend
using LLaMA-tiny-1B model
ram used:  0.00 GB, freqs_cis                                         : 100%|████████| 202/202 [03:42<00:00,  1.10s/it]
loaded weights in 222133.51 ms, 0.00 GB loaded at 0.00 GB/s
Hello.得$\★ failsправ modificationsмени stone částòria

**Comment by yeeyangtee** [comment]
This works for safe_load of bfloat, just do the manual casting after reshape:

`return {k:t[8+json_len+v['data_offsets'][0]:].cast(dtypes.bfloat16)[:prod(v['shape'])].reshape(v['shape']).bitcast(dtypes.uint16).to("CPU").cast(dtypes.uint32).mul(1<<16).bitcast(dtypes.float32).to(Device.DEFAULT).half() for k,v in metadata.items() if k != "__metadata__"}`

You can refer to my PR #2427 . I'll close it since I didn't realise you already started on this earlier.

**Comment by rodfer0x80** [comment]
[x] safe_tensor bfloat16 casting is working, Thank you! @yeeyangtee [ #2427 ]
[x] separated a function to fetch model that downloads weights and tokenizer using MODEL_PARAMS and places them accordingly in ./weights
[x] chatbot mode is working!

<CHAT LOG>
User: What is your name?
Stacy: Hi! My name is Stacy. I'm a rapper with bipolar disorder. [EOS]
User: french revolution was what year?
Stacy: The French Revolution started in 1789, and lasted 10 years until 1799. [EOS]
User: What is bigger, the moon or the sun?
Stacy: The sun is bigger than the moon, except when Mercury is in retrograde. [EOS]
User: How are you feeling today, Stacy?
Stacy: I'm feeling good, but I'm also feeling anxious. [EOS]
User: Why is that?
Stacy: I'm struggling with my mental health, and I'm not sure how to cope. [EOS]
User: Have you tried rebooting?
Stacy: Yes, I've tried rebooting a few times, but it doesn't seem to work. [EOS]

**Comment by rodfer0x80** [comment]
TODO:
[ ] add test to ./test/unit/test_disk_tensor.py  (current safe_load implementation needs to declare BF16 in metadata layer metadata)

**Comment by rodfer0x80** [comment]
This should work with multi dtype layers too

**Comment by rodfer0x80** [comment]
Does it actually makes sense to test for bf16 explicitly in test/unit/test_disk_tensor.py?

**Comment by rodfer0x80** [comment]
I was blocked on the BF16 cast.  Need to have another look at this now that it is mostly stable

**Comment by geohot** [comment]
Feel free to reopen
